test/mobile/custom_build/prepare_model.py

```diff
@@ -13,5 +13,6 @@
 
 # Dump root ops used by the model (for custom build optimization).
 ops = torch.jit.export_opnames(traced_script_module)
+ops.append('aten::ones')  # HACK because predictor.cpp explicitly calls this!
```

Comment:
Just a thought that if our eventual setup still needs to explicitly push roots into `ops` like this we'll probably want to institutionalize it, like have a global list called `EXTRA_ROOT_OPS` or whatever

Reply:
Yes, as I replied to iseeyuan, we need run static analysis against JIT code to find out common EXTRA_ROOT_OPS, if any.

However, this case is different - aten::ones is directly called from *client* code (the dummy predictor.cpp calls it to create all-one-tensor for testing purpose). Generally, user could write pure c++ client code to directly call functions or tensor methods, which won't be captured by extracting ops from model's bytecodes.

This is less a problem for Android, where we expect users to use the limited set of Java APIs. To actually solve this problem, we probably need ask users to run code analyzer against their client code to dump these extra root ops - that's why I feel there will be more support work to switch to dynamic dispatch - in static dispatch case these extra ops will be kept by linker automatically.

For CI purpose this one-off hack is probably fine? :)

Reply:
Question, does this mean every custom build is gonna get `aten::ones` whether they use it or not? Or just binaries that use the dummy `predictor.cpp`? (I'm guessing the former since this is a general script, but plz correct :) 

A single op isn't that big a deal I guess, but I think it's important to describe the situation more explicitly - something like your description above in a comment above a global that declares the op more explicitly. Text below is just your comment above, so may not be entirely correct, but this is the kind of thing I mean:

```
# We need to run static analysis against JIT code to find out common EXTRA_ROOT_OPS, if any.
# 
# However, this case is different - aten::ones is directly called from client code (the dummy
# predictor.cpp calls it to create all-one-tensor for testing purpose). Generally, user could
# write pure c++ client code to directly call functions or tensor methods, which won't be
# captured by extracting ops from model's bytecodes.
# 
# This is less a problem for Android, where we expect users to use the limited set of Java 
# APIs. To actually solve this problem, we probably need ask users to run code analyzer against 
# their client code to dump these extra root ops - that's why I feel there will be more support 
# work to switch to dynamic dispatch - in static dispatch case these extra ops will be kept by 
# linker automatically.
# 
# For CI purpose this one-off hack is probably fine? :)
EXTRA_CI_ROOT_OPS = ['aten::ones']

...
ops.extend(EXTRA_CI_ROOT_OPS)
```

Reply:
> Question, does this mean every custom build is gonna get `aten::ones` whether they use it or not? Or just binaries that use the dummy `predictor.cpp`? (I'm guessing the former since this is a general script, but plz correct :)

It's the latter - the script is under test/mobile/... which is for mobile CI specifically. That's why I don't feel so guilty to add this hack :)

Reply:
Added the comment to the test script to keep a record.